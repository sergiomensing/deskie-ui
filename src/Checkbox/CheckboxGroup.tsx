import { PropsWithChildren, HTMLAttributes } from 'react';
import styled, { css } from 'styled-components';

type CheckboxGroupProps = HTMLAttributes<unknown> & { direction?: 'vertical' | 'horizontal' };

const StyledCheckboxGroup = styled.div<{ horizontal: boolean }>`
  display: flex;

  ${(props) =>
    props.horizontal
      ? css`
          align-items: center;
          flex-wrap: wrap;
          margin: -0.5rem -0.75rem;

          & > * {
            padding: 0.5rem 0.75rem;
          }
        `
      : css`
          flex-direction: column;

          & > * + * {
            margin-top: 0.5rem;
          }
        `}
`;

function CheckboxGroup({
  children,
  direction = 'horizontal',
  className = '',
  ...rest
}: PropsWithChildren<CheckboxGroupProps>) {
  return (
    <StyledCheckboxGroup horizontal={direction === 'horizontal'} className={`checkbox-group ${className}`} {...rest}>
      {children}
    </StyledCheckboxGroup>
  );
}

export default CheckboxGroup;
