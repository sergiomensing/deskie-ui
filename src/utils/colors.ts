export const colorShade = (col: string, amt: number) => {
  col = col.replace(/^#/, '');
  if (col.length === 3) col = col[0] + col[0] + col[1] + col[1] + col[2] + col[2];

  const [sr, sg, sb] = col.match(/.{2}/g) || ['ff', 'ff', 'ff'];
  const [ar, ag, ab] = [parseInt(sr, 16) + amt, parseInt(sg, 16) + amt, parseInt(sb, 16) + amt];

  const r = Math.max(Math.min(255, ar), 0).toString(16);
  const g = Math.max(Math.min(255, ag), 0).toString(16);
  const b = Math.max(Math.min(255, ab), 0).toString(16);

  const rr = (r.length < 2 ? '0' : '') + r;
  const gg = (g.length < 2 ? '0' : '') + g;
  const bb = (b.length < 2 ? '0' : '') + b;

  return `#${rr}${gg}${bb}`;
};

const defaultColors = [
  '#E64980',
  '#4C6EF5',
  '#228BE6',
  '#15AABF',
  '#12B886',
  '#40C057',
  '#82C91E',
  '#FAB005',
  '#FD7E14',
  '#FA5252',
];

export function getColorFromString(value: string, colors: undefined | string[] = defaultColors) {
  let hash = 0;
  if (value.length === 0) return colors[hash];
  for (let i = 0; i < value.length; i++) {
    hash = value.charCodeAt(i) + ((hash << 5) - hash);
    hash = hash & hash;
  }
  hash = ((hash % colors.length) + colors.length) % colors.length;
  return colors[hash];
}
