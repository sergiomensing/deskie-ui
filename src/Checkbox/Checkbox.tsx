import {
  forwardRef,
  InputHTMLAttributes,
  PropsWithChildren,
  useImperativeHandle,
  useRef,
  useState,
  useEffect,
  useCallback,
} from 'react';
import CheckboxGroup from './CheckboxGroup';
import styled, { css } from 'styled-components';

type BaseCheckboxProps = {
  onChange?: (ev: React.ChangeEvent<HTMLInputElement>) => void;
};

export type CheckboxProps = BaseCheckboxProps & Omit<InputHTMLAttributes<unknown>, keyof BaseCheckboxProps | 'type'>;

const StyledCheckbox = styled.label<{ disabled: boolean; checked: boolean }>`
  cursor: ${(props) => (props.disabled ? 'not-allowed' : 'pointer')};
  display: flex;
  align-items: flex-start;
  user-select: none;
  color: ${(props) => (props.disabled ? props.theme.colors.disabledText : props.theme.colors.label)};

  input[type='checkbox'] {
    opacity: 0;
    outline: none;
    position: absolute;
    width: 0;
    height: 0;
    margin: 0;
    padding: 0;
    z-index: -1;
    background-color: transparent;
  }

  .checkbox-icon {
    display: flex;
    align-items: center;
    margin-right: 0.5rem;
  }

  .checkbox-icon:before {
    content: '';
    width: 1.25rem;
    height: 1.25rem;
    border-radius: 0.25rem;
    display: flex;
    justify-content: center;
    align-items: center;
    border: 2px solid ${(props) => (props.disabled ? props.theme.colors.disabled : props.theme.colors.border)};
    transition: 0.1s ease;
  }

  .checkbox-icon:after {
    content: '';
    opacity: 0;
    background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' height='24' width='24'%3E%3Cpath d='M0 0h24v24H0z' fill='none' /%3E%3Cpath fill='%23ffffff' d='M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z'/%3E%3C/svg%3E");
    background-size: contain;
    background-position: center;
    background-repeat: no-repeat;
    width: 1.25rem;
    height: 1.25rem;
    border-radius: 0.25rem;
    position: absolute;
    border: 2px solid transparent;
  }

  .checkbox-label {
    margin-top: 2px;
    white-space: nowrap;
  }

  input:checked ~ .checkbox-icon::before {
    background-color: ${(props) => (props.disabled ? props.theme.colors.disabled : props.theme.colors.primary)};
    border-color: ${(props) => (props.disabled ? props.theme.colors.disabled : props.theme.colors.primary)};
  }

  ${(props) =>
    !props.disabled &&
    css`
      &:hover .checkbox-icon::before {
        border-color: ${(props) => props.theme.colors.inputBorder};
      }
      &:hover input:checked ~ .checkbox-icon::before {
        background-color: ${(props) => props.theme.colors.primaryHover};
        border-color: ${(props) => props.theme.colors.primaryHover};
      }

      input:focus-visible ~ .checkbox-icon:before {
        box-shadow: 0 0 0 2px ${(props) => props.theme.colors.inputBorder};
      }
      input:checked:focus-visible ~ .checkbox-icon::before {
        box-shadow: 0 0 0 2px ${(props) => props.theme.colors.primaryHover};
      }
    `}

  input:checked ~ .checkbox-icon::after {
    opacity: 1;
  }
`;

const Checkbox = forwardRef<HTMLInputElement, PropsWithChildren<CheckboxProps>>(
  (
    { disabled = false, checked, defaultChecked, children, onChange, className = '', ...rest },
    ref: React.Ref<HTMLInputElement | null>,
  ) => {
    const [selfChecked, setSelfChecked] = useState<boolean>(() => defaultChecked || false);

    const changeHandle = useCallback(
      (ev: React.ChangeEvent<HTMLInputElement>) => {
        if (disabled) {
          return;
        }

        setSelfChecked(!selfChecked);

        if (onChange) {
          const event: React.ChangeEvent<HTMLInputElement> = { ...ev, target: { ...ev.target, checked: !selfChecked } };
          onChange(event);
        }
      },
      [onChange, disabled, selfChecked],
    );

    useEffect(() => {
      if (checked === undefined) {
        return;
      }
      setSelfChecked(Boolean(checked));
    }, [checked]);

    const checkboxRef = useRef<HTMLInputElement>(null);
    useImperativeHandle(ref, () => checkboxRef.current);

    return (
      <StyledCheckbox className={`checkbox-container ${className}`} disabled={disabled} checked={selfChecked}>
        <input
          type="checkbox"
          ref={checkboxRef}
          onChange={changeHandle}
          checked={selfChecked}
          disabled={disabled}
          {...rest}
        />
        <span className="checkbox-icon"></span>
        {children && <span className="checkbox-label">{children}</span>}
      </StyledCheckbox>
    );
  },
);

export default Checkbox as typeof Checkbox & { Group: typeof CheckboxGroup };
