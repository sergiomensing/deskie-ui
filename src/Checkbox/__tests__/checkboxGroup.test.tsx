import React from 'react';
import Checkbox from '../Checkbox';
import { render } from '../../../tests/utils';

describe('Checkbox Group', () => {
  it('should support direction', () => {
    const { container } = render(
      <>
        <Checkbox.Group direction="horizontal">
          <Checkbox checked={true}>test-value</Checkbox>
          <Checkbox checked={false}>test-value</Checkbox>
        </Checkbox.Group>
        <Checkbox.Group direction="vertical">
          <Checkbox checked={true}>test-value</Checkbox>
          <Checkbox checked={false}>test-value</Checkbox>
        </Checkbox.Group>
      </>,
    );
    expect(container).toMatchSnapshot();
  });
});
