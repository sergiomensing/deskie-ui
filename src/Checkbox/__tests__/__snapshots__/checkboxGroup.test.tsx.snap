// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Group should support direction 1`] = `
.c1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #11142D;
}

.c1 input[type='checkbox'] {
  opacity: 0;
  outline: none;
  position: absolute;
  width: 0;
  height: 0;
  margin: 0;
  padding: 0;
  z-index: -1;
  background-color: transparent;
}

.c1 .checkbox-icon {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 0.5rem;
}

.c1 .checkbox-icon:before {
  content: '';
  width: 1.25rem;
  height: 1.25rem;
  border-radius: 0.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 2px solid #D3D4DD;
  -webkit-transition: 0.1s ease;
  transition: 0.1s ease;
}

.c1 .checkbox-icon:after {
  content: '';
  opacity: 0;
  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' height='24' width='24'%3E%3Cpath d='M0 0h24v24H0z' fill='none' /%3E%3Cpath fill='%23ffffff' d='M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z'/%3E%3C/svg%3E");
  background-size: contain;
  background-position: center;
  background-repeat: no-repeat;
  width: 1.25rem;
  height: 1.25rem;
  border-radius: 0.25rem;
  position: absolute;
  border: 2px solid transparent;
}

.c1 .checkbox-label {
  margin-top: 2px;
  white-space: nowrap;
}

.c1 input:checked ~ .checkbox-icon::before {
  background-color: #6C5DD3;
  border-color: #6C5DD3;
}

.c1:hover .checkbox-icon::before {
  border-color: #9395A1;
}

.c1:hover input:checked ~ .checkbox-icon::before {
  background-color: #4535B7;
  border-color: #4535B7;
}

.c1 input:focus-visible ~ .checkbox-icon:before {
  box-shadow: 0 0 0 2px #9395A1;
}

.c1 input:checked:focus-visible ~ .checkbox-icon::before {
  box-shadow: 0 0 0 2px #4535B7;
}

.c1 input:checked ~ .checkbox-icon::after {
  opacity: 1;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin: -0.5rem -0.75rem;
}

.c0 > * {
  padding: 0.5rem 0.75rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 > * + * {
  margin-top: 0.5rem;
}

<div>
  <div
    class="c0 checkbox-group "
  >
    <label
      class="c1 checkbox-container "
    >
      <input
        type="checkbox"
      />
      <span
        class="checkbox-icon"
      />
      <span
        class="checkbox-label"
      >
        test-value
      </span>
    </label>
    <label
      class="c1 checkbox-container "
    >
      <input
        type="checkbox"
      />
      <span
        class="checkbox-icon"
      />
      <span
        class="checkbox-label"
      >
        test-value
      </span>
    </label>
  </div>
  <div
    class="c2 checkbox-group "
  >
    <label
      class="c1 checkbox-container "
    >
      <input
        type="checkbox"
      />
      <span
        class="checkbox-icon"
      />
      <span
        class="checkbox-label"
      >
        test-value
      </span>
    </label>
    <label
      class="c1 checkbox-container "
    >
      <input
        type="checkbox"
      />
      <span
        class="checkbox-icon"
      />
      <span
        class="checkbox-label"
      >
        test-value
      </span>
    </label>
  </div>
</div>
`;
